### DISCUSSION TITLE
Websocket Route Not Working With Nested Router

### DISCUSSION THREAD

**Discussion by uberbrodt** [fastapi/fastapi#6404] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter, FastAPI
from fastapi import WebSocket


router = APIRouter(prefix="/foo")
nested_router = APIRouter()
router.include_router(nested_router)
app = FastAPI()


# WORKS

# python -m websockets ws://localhost:8000/test-ws
@app.websocket("/test-ws")
async def test_ws(ws: WebSocket):
    await ws.accept()
    await ws.send_text("Hello World!")
    await ws.close()


# WORKS

# python -m websockets ws://localhost:8000/foo/test-ws
@router.websocket("/test-ws")
async def foo_test_ws(ws: WebSocket):
    await ws.accept()
    await ws.send_text("Hello World!")
    await ws.close()


# FAILS - 403 FORBIDDEN

# python -m websockets ws://localhost:8000/foo/bar/test-ws
@nested_router.websocket("/bar/test-ws")
async def bar_test_ws(ws: WebSocket):
    await ws.accept()
    await ws.send_text("Hello World!")
    await ws.close()


app.include_router(router)
```

### Description

When upgrading from FastaAPI 0.72.0 to 0.78.0, noticed that a websocket path I had defined on a router included by _another_ router was no longer available (server returns 403, trace logs indicate normal closure, think this is the normal behaviour when a websocket tries to upgrade a non-existent route).

Looks like #2640 fixed the case for a websocket defined on a router, but seems like the nested router makes this appear again.

### Operating System

macOS

### Operating System Details

<img width="448" alt="Screen Shot 2022-05-19 at 10 47 43 AM" src="https://user-images.githubusercontent.com/978339/169342214-fa0aeeb5-815f-4c1c-9785-c36d8dac19c8.png">

### FastAPI Version

0.78.8

### Python Version

3.10.1

### Additional Context

It's kinda weird that it _did work_ before, since #98 would indicate that it's been a long standing problem. 🤷

**Comment by uberbrodt** [comment]
Alright so I pulled down the code and it seems like I can fix it with ordering the path definitions and the include_router. If I restructure the code like this:

```
from fastapi import APIRouter, FastAPI
from fastapi import WebSocket

nested_router = APIRouter()

# NOW WORKS 
# python -m websockets ws://localhost:8000/foo/bar/test-ws
@nested_router.websocket("/bar/test-ws")
async def bar_test_ws(ws: WebSocket):
    await ws.accept()
    await ws.send_text("Hello World!")
    await ws.close()

router = APIRouter(prefix="/foo")

# WORKS
# python -m websockets ws://localhost:8000/foo/test-ws
@router.websocket("/test-ws")
async def foo_test_ws(ws: WebSocket):
    await ws.accept()
    await ws.send_text("Hello World!")
    await ws.close()

router.include_router(nested_router)
app = FastAPI()

# WORKS
# python -m websockets ws://localhost:8000/test-ws
@app.websocket("/test-ws")
async def test_ws(ws: WebSocket):
    await ws.accept()
    await ws.send_text("Hello World!")
    await ws.close()

app.include_router(router)
```

I can then hit the  nested router:
```
cbrodt@cbrodt-safely-you[fastapi]$  python -m websockets ws://localhost:8000/foo/bar/test-ws
Connected to ws://localhost:8000/foo/bar/test-ws.
< Hello World!
Connection closed: 1000 (OK).
```

So it seems like when calling include_router, it will miss any path operations not yet defined. Not sure on the right approach for a fix, but in the meantime maybe add a warning to the docs?

**Comment by qzcloveswcy** [comment]
Hello, I recently tried how to initiate a request to websocket in the sub-application, but found that the websocket parameter cannot be received in the sub-route. How to solve it?

**Comment by Kludex** [comment, marked as answer]
- Solved by https://github.com/tiangolo/fastapi/pull/2640

**Comment by ChrisNi888** [comment]
While learning FastAPI WebSockets, I'm using the encode/broadcaster (https://github.com/encode/broadcaster) as mentioned in the tutorial. I've found that my client can send messages to the server, but the server seems to not have successfully subscribed and does not receive messages from the subscribed channel.

I've checked this issue for a long time, and I've found that if I write it directly in the app's path, like this:


`@app.websocket("/chat_room")`
the result is normal.
However, if I create a router, like this:


```
router = APIRouter(
    prefix="/ws",
    tags=["WebSocket"],
)

@router.websocket("/chat_room")
```
the problem mentioned above occurs.
The reason is unclear.
